31 - Create the form elements of the design system

MellyGray opened the pull request:
## What this PR does / why we need it:
This PR add the  form elements of the design system as described in the [Design System Analysis Doc](https://docs.google.com/document/d/1OxCQO7B-yJaMWRl8EzuflcSmIaCmVSs19aoD83PjPmA/edit#heading=h.8nqu9gh25f0n)


## Which issue(s) this PR closes:

- Closes #31 

## Special notes for your reviewer:
I did't add all the form elements in the Dataverse because that could be infinite, I only added the ones we'll be using in the MVP.

For the tests I didn't test some FormElements independently because it didn't make sense to for example test the input component without the label component.

## Suggestions on how to test this:

npm run storybook

Go to storybook and check all the different form elements in the Form doc file. You can even see the code from the storybook by clicking the code button in each story.

## Does this PR introduce a user interface change? If mockups are available, please link/include them here:
None

## Is there a release notes update needed for this change?:
Form elements of the design system added

## Additional documentation:
https://docs.google.com/document/d/1OxCQO7B-yJaMWRl8EzuflcSmIaCmVSs19aoD83PjPmA/edit#heading=h.8nqu9gh25f0n

pdurbin commented on src/stories/ui/form/Form.stories.tsx:
```diff
@@ -0,0 +1,252 @@
+export const GroupWithMultipleFields: Story = {
+  render: () => (
+    <Form>
+      <Form.GroupWithMultipleFields title="Author">
+        <Row>
+          <Form.Group as={Col} controlId="basic-form-name">
+            <Form.Group.Label>Name</Form.Group.Label>
+            <Form.Group.Input type="text" placeholder="Name" />
+            <Form.Group.Text>
+              Please, specify the authors name. If there are multiple authors, please separate them
+              with a comma.
+            </Form.Group.Text>
+          </Form.Group>
+          <Form.Group as={Col} controlId="basic-form-surname">
+            <Form.Group.Label>Surname</Form.Group.Label>
+            <Form.Group.Input type="text" placeholder="Surname" />
+          </Form.Group>
+        </Row>
```
I'm not sure how accurate this is supposed to be but the way authors are presented is pretty different than what we do in reality. Maybe we should switch to a different example? Or make it more real? Maybe it's just a matter or changing Surname to Affiliation?

To add multiple authors, we use a plus sign to add more, like this:

![Screen Shot 2023-04-24 at 4 42 53 PM](https://user-images.githubusercontent.com/21006/234112732-efd76596-e3ff-4eca-bb89-aa6abafa1a4c.png)

Also, authors just have a single "Name" field with can be a person or an organization. So it's a little weird to see Name and Surname here (for user accounts we use Given Name and Family Name, by the way:

![Screen Shot 2023-04-24 at 4 42 19 PM](https://user-images.githubusercontent.com/21006/234112975-8b6cd25c-5379-4ac9-aab3-b1fafcd7e1b7.png)

MellyGray replied:
When I did this I was just thinking of an example, not necessary a real one

Maybe I can think of a different example not related to Dataverse, because I would like to have all the different input types in one example

About the plus sign, should I add it to this PR as part of the form elements?

pdurbin replied:
I do think the plus sign is important to have somewhere. Yes, if you could add it, that would be great.

I'm happy to help make the examples a little more real. Otherwise, it's a bit confusing.

MellyGray replied:
I added the the plus button functionality and fixed the examples, let me know if I need to change something else

pdurbin commented on src/stories/ui/form/Form.stories.tsx:
```diff
@@ -0,0 +1,252 @@
+
+/**
+ * ## Description
+ * A form is a collection of HTML elements used to gather user input. Allows users to enter data, such as text, numbers,
```
```suggestion
 * A form is a collection of HTML elements used to gather user input. It allows users to enter data, such as text, numbers,
```

pdurbin commented on src/stories/ui/form/Form.stories.tsx:
```diff
@@ -0,0 +1,252 @@
+ *  - Add the text below the input
+ * - Select:
+ *  - First option should be the 'Select...' option
+ *  - The options should use the <option> tag
```
`<option>` doesn't show up. Put it in backticks, maybe?

MellyGray replied:
fixed!

pdurbin commented on src/stories/ui/form/Form.stories.tsx:
```diff
@@ -0,0 +1,252 @@
+        </Row>
+        <Row>
+          <Form.Group as={Col} controlId="basic-form-description">
+            <Form.Group.Label>Description</Form.Group.Label>
```
Authors don't have descriptions in Dataverse so this example feels weird.

MellyGray replied:
Yes, sorry, I just wanted to show all the different input types in one example. I can think of a different example not related with Dataverse

MellyGray replied:
Ok, so I changed the examples to these ones which come from the New Dataset form. Let my know if they make more sense know. I'm working on the plus button to add it to the examples too.

![image](https://user-images.githubusercontent.com/23359572/234582646-38f4dfee-f765-4b56-b9a4-f396f44f5fed.png)

pdurbin reviewed (commented):
Overall this looks good. I did add some suggestions.

pdurbin reviewed (approved):
The examples look great. I like the plus button and it even works. Thanks! Approved!

melinahernandeztheam commented:
> Oh, the plus sign does look a little different...
> 
> 
> 
> ## old, JSF
> 
> 
> 
> ![234112732-efd76596-e3ff-4eca-bb89-aa6abafa1a4c](https://user-images.githubusercontent.com/21006/235194956-bfbdc71c-fd2f-41b4-9b5a-94517c288f49.png)
> 
> 
> 
> ## new, React
> 
> 
> 
> ![Screen Shot 2023-04-28 at 11 50 44 AM](https://user-images.githubusercontent.com/21006/235194931-71d7377a-9eef-4b2a-a3e7-7f8658bf0b3f.png)
> 
> 
> 
> ... but I guess that's ok. I don't want to hold this up.
> 
> 
> 
> I'm also not sure about the ? and "Add" tooltip... if we should be expecting them or not. This tooltip PR was merged already:
> 
> 
> 
> - #55

You're right, I can fix the plus button and add the Add tooltip. Or if we don't want to hold this, then we can create a separated issue for integrating the tooltips with all the form inputs components and to fix the plus button. Both ways work for me

pdurbin commented:
I just asked @ekraffmiller about this in Slack as well.

Being new to Storybook, I guess I'm not 100% clear on how realistic the stories should be. I think I like the ideas of taking the smaller components like tooltips and adding them to forms (where they are often used). However, I'm not trying to create more work for us. Once we get to pages, I think the idea is that they should look fairly realistic. They should probably have tooltips, for example. Again, I'm not sure. 🤔 

Oh, and by the way, I pushed a little commit: 656b4d2

No QA today but if we decide to work on this more, we should pull it out of QA. I'm happy to go with the majority opinion. 😄

ekraffmiller commented:
Is the idea to add the question tooltip as a part of the FormLabel component, like the RequiredInputSymbol?

pdurbin commented:
Yes. Something like this is what I had in mind:

![Screen Shot 2023-04-28 at 2 09 20 PM](https://user-images.githubusercontent.com/21006/235222110-1afb0c2b-7160-4682-9240-b52957640891.png)

This is only one of the five tooltips, of course. And I'm not sure of the best way to add more space between "Affiliation" and the tooltip. 😅 

This is what I'm doing:

```
$ git diff
diff --git a/src/stories/ui/form/Form.stories.tsx b/src/stories/ui/form/Form.stories.tsx
index 254b222..fc521b8 100644
--- a/src/stories/ui/form/Form.stories.tsx
+++ b/src/stories/ui/form/Form.stories.tsx
@@ -2,6 +2,7 @@ import type { Meta, StoryObj } from '@storybook/react'
 import { Form } from '../../../sections/ui/form/Form'
 import { Col } from '../../../sections/ui/grid/Col'
 import { Row } from '../../../sections/ui/grid/Row'
+import { Tooltip } from '../../../sections/ui/tooltip/Tooltip'
 
 /**
  * ## Description
@@ -210,6 +211,7 @@ export const FormValidation: Story = {
           </Form.Group>
           <Form.Group as={Col} controlId="basic-form-affiliation">
             <Form.Group.Label>Affiliation</Form.Group.Label>
+            <Tooltip placement="auto" message="The name of the entity affiliated with the author, e.g. an organization's name"></Tooltip>
             <Form.Group.Input type="text" placeholder="Affiliation" />
           </Form.Group>
         </Row>
```

Anyway, these tool tips are all over our forms so maybe we should go ahead and add them in this PR? I'm going to pull this out of "Ready for QA" (plenty of other stuff in the currently anyway), so we can decide how to proceed. Besides, it sounds like @MellyGray wants to add the plus button. I'll assign all three of us but please unassign yourself if you want off! 😄

MellyGray commented:
@ekraffmiller @pdurbin I added the tooltips and fixed the plus button 

![image](https://user-images.githubusercontent.com/23359572/235659985-d0861bf3-24ed-4c79-bfbc-495e75e5d3b9.png)

The plus button doesn't look exactly the same because this is how the Plus icon looks in Bootstrap 5:

![Captura de pantalla 2023-05-02 a las 14 00 07](https://user-images.githubusercontent.com/23359572/235660418-534345be-da21-40eb-bc00-99341b40231d.png)

Let me know what you think

ekraffmiller reviewed (approved):
Looks good to me, thanks!

kcondon commented:
@MellyGray This looks good but there is a merge conflict. Would you please resolve this?  Btw, on some of the Props, Group With Multiple Fields, when you click + or -, it does add a new instance but it also clears any existing instances of data. Is that something we would also see when on the page because that might cause some issues if people have not saved the page yet and wanted to add another entry. If this iteration is to get it roughly correct and adjust later, then np. Thanks!
